design/reboot-interface.md

```diff
@@ -0,0 +1,155 @@
+reboots to be manipulated or cancelled, and a record to be left behind of past
+scheduled reboots. Any such resource would have the same requirements of the
+BareMetalHost as fencing does.
+
+This API is not responsible for managing unprovisioned hosts, e.g. to recover
+from state inconsistencies in Ironic prior to provisioning. Any such
+inconsistencies that are not handled internally by the BareMetalHost will
+likely need manual intervention to recover anyway.
+
+## Proposal
+
+A new date-time field, ``lastBooted``, will be added to the ``provisioning``
+section of the BareMetalHost status. This records a time after which the Host
+was last booted using the current image. Processes running prior to this time
+may be assumed to have been fenced.
+
+To provide a declarative interface for performing a reboot, a new
+``rebootRequestTimestamp`` field will to be added to the BareMetalHost spec.
+This field could have any type, so long as clients are able to provide a unique
+value to trigger another reboot, but the current timestamp at the time of the
+request is the logical value to use.
+
+Request timestamps that predate the last booted time will be ignored. Request
+timestamps after the last booted time must be honoured, regardless of the order
+in which they were issued. This eliminates the need for clients to implement
+logic to determine if/when they should overwrite an existing reboot request.
+
+All reboot requests between the last booted time and the present can be
+condensed into a single reboot, since this will necessarily move the last
+booted time to a point after all of the reboot requests. The same is not true
+for timestamps in the future (whether due to clock skew or a misguided attempt
+to use this interface to schedule future reboots), since the new last booted
+time may still fall before some portion of the future-dated requests. Such
+requests will be treated as if they occurred at the present time rather than at
+the future time given in the timestamp.
+
+A new ``pendingRebootSince`` field in the status will contain the timestamp of
+the earliest reboot request following the last booted time, if there has been
+any and the machine is provisioned.
+
+Client code may be assured that the node has been fenced when any of the
+following are true:
+
+* The Host has been deprovisioned (in this case the ``lastBooted`` field will
+  be empty)
+* The ``lastBooted`` time is after the time of the reboot request
+* The ``poweredOn`` status is ``false``
+* The ``pendingRebootSince`` time is empty (checking this is necessary only if
+  fairly severe clock skew is possible)
+
+## Drawbacks
+
+The need for the consumer to supply a recent timestamp means that the design
+requires some co-operation from the client that is not naturally enforced by
+the API. Without good documentation this could lead to confusion and incorrect
+usage. It is also vulnerable to clock skew in severe cases.
+
+## Alternatives
```

Comment:
Here's an option that's growing on me and that I think is worth serious consideration. It takes inspiration from how finalizers work. @jarrpa this is relevant to your use case in ocs-operator; it's another option for inducing "factory reset" of your data.

In the host's Spec, we add a field called `rebootRequests`. It is an array of strings. If you want a reboot, you put a string of your choice in that array.

When the controller sees one or more strings in the array, it begins a reboot.

If additional strings are added while the reboot is in-progress, the controller ignores them.

When the reboot is complete, the controller clears the array.

It allows multiple actors to safely request reboots at close to the same time. It allows each actor to know if its reboot request has been completed. And it enables the controller to know for sure if anyone has requested a reboot since the previous reboot.

The only part that's a bit unconventional is that the controller is modifying the spec for its primary resource. But in this case I think that remains within the spirit of the declarative API.

To get around that, we could theoretically make a separate controller that watches for entries in `status.rebootRequestsCompleted` and deletes them from `spec.rebootRequests`, and then have our main controller remove those requests from `status.rebootRequestsCompleted` as they disappear from the spec. That would fit the pattern of only writing to your primary resource's status and letting others write to your spec, but it would be more code and complexity for no practical gain. It's easier to let the one control loop just clear out the entries from the spec; the API experience is the same ether way for the actors who want reboots.

WDYT?

Reply:
This document asserts that we need an atomic reboot operation, but doesn't really explain why. I would like to understand exactly where the race condition (or other issue) is in the design that just uses the existing power switch with a separate CRD to indicate the request for a reboot. I went through a few cases in another comment and only came up with one race (the back-to-back reboot), which wouldn't be solved by any of the proposed approaches and which I don't consider all that bad (we're not going to impose a minimum-time-between-reboots, are we?).

Is there actually a race condition if we have a properly written reboot controller that examines the host state before deciding what action to take?

Reply:
@dhellmann I would say that first and foremost, having a declarative interface is its own reward. You certainly can work around design failures by sequentially updating declarative interfaces, but it gets un-fun pretty quickly and a system where such hacks are allowed to proliferate freely turns into a big ball of mud. As far as specific races go, there's the case where an admin attempts to power off the server during the period where the fencing code is trying to reboot it, which would result in the server being unintentionally booted again. In the opposite case, where an operator exists that is forcing the power state to be on, we might never be able to fence the node (the CAPBM doesn't currently do this, but if we can't convince upstream cluster-api to add a reboot interface then our Machine reboot operator should be completely independent of CAPBM).

@mhrivnak my reservation about doing it as a list in the main Host spec is that IIUC it pushes some of the responsibility for correctness onto the clients - a client wanting to request a reboot must append to the list without removing any other entries, otherwise the source of the other entries could legitimately conclude that its reboot request had been completed. Copying entries to the status as they're received could help, but I think there's still a race. We shouldn't rely on clients being well-behaved in terms of implementing logic correctly to ensure they play nice with others. ``kubectl apply`` has no such logic, for a start.

Reply:
What about having the `rebootRequestTimestamp` field, but having the BMH controller create a corresponding `HostRebootRequest`, or update it if it already exists (depending on how the value of `rebootRequestTimestamp` changes)?

Reply:
@zaneb re: your concern about correct array handling, that's already how finalizers work. We depend on clients to correctly add and remove their finalizer without affecting other finalizers in the array. It's simple, safe and common.

Reply:
I don't see anything in the design that addresses the issue of conflicting requests to reboot and set the power state -- they're separate fields and can be changed in separate calls (or the same call) in ways that conflict.

If we want to prevent simultaneous action, then either we need to use a web hook to validate the inputs or we need to redesign the power interface at the same time to combine the two things.

That said, I don't feel like I've seen a lot of other kubernetes APIs that worried about conflicting requests. Isn't the point that the operator reconciles the state of the world to match the resource based on its most current values precisely so it isn't necessary to design APIs that understand all of the transitions between different states?

Reply:
The design makes the reboot request independent of the long-term desired power state. So `oc patch -type patch {"spec": {"online": false}}` always results in the power remaining off, whereas if the fencing code were twiddling the power state around the same time then the power might come back on even after issuing that command.
Generally that should only happen in the case where it's controlled by another operator, and the other operator provides the preferred user interface for setting the long-term desired state. That isn't the case here, because there isn't another interface for controlling the power state; users should expect that things they set here will stick.